src/drivers/clock/epit/epit.c

```diff
@@ -0,0 +1,99 @@
+/**
+ * @file epit.c
+ * @brief Enhanced.Periodic.Interrupt.Timer driver
+ * @author Puranjay Mohan <puranjay12@gmail.com>
+ * @version 0.1
+ * @date 2020-03-05
+ */
+
+#include <drivers/common/memory.h>
+#include <hal/clock.h>
+#include <hal/reg.h>
+#include <kernel/irq.h>
+#include <kernel/printk.h>
+#include <kernel/time/clock_source.h>
+#include <kernel/time/time_device.h>
+
+#include <framework/mod/options.h>
+#include <embox/unit.h>
+
+#define EPIT_BASE       OPTION_GET(NUMBER, base_addr)
+#define EPIT_IRQ        OPTION_GET(NUMBER, irq_nr)
+#define EPIT_TARGET_HZ  1000
+
+#define EPIT_CR 		(EPIT_BASE + 0x00) /* Control Register */
+#define EPIT_SR     	(EPIT_BASE + 0x04) /* Status Register */
+#define EPIT_LR     	(EPIT_BASE + 0x08) /* Load Register */
+#define EPIT_CMPR   	(EPIT_BASE + 0x0C) /* Compare Register */
+#define EPIT_CNR    	(EPIT_BASE + 0x10) /* Counter Register */
+
+#define EPIT_CR_PRESCLR (0x42 << 4)/* Prescaler value of 66 */
+#define EPIT_CR_EN 		(1 << 0) /* Enable Bit */
+#define EPIT_CR_ENMOD   (1 << 1) /* Enable Mode Bit */
+#define EPIT_CR_OCIEN   (1 << 2) /* Output compare Interrupt Enable bit*/
+#define EPIT_CR_RLD		(1 << 3) /* Counter Reload Control bit */
+#define EPIT_SR_CLEAR	(1 << 0) /* Clear the status register*/
+#define EPIT_CR_CLKSRC	(1 << 24) /* Use peripheral clock as input, 66MHz */
+#define EPIT_CR_IOVW	(1 << 17) /* Load Immidiate Override*/
+
+static struct clock_source epit_clock_source;
+static irq_return_t clock_handler(unsigned int irq_nr, void *data) {
+	clock_tick_handler(irq_nr, data);
+	REG32_STORE(EPIT_SR, EPIT_SR_CLEAR);
+	return IRQ_HANDLED;
+}
+
+static int epit_init(void) {
+	clock_source_register(&epit_clock_source);
+
+	return irq_attach(EPIT_IRQ,
+	                  clock_handler,
+	                  0,
+	                  &epit_clock_source,
+	                  "EPIT");
+}
+
+static int epit_config(struct time_dev_conf * conf) {
+	/* changin clock source to peripheral clock */
+	REG32_STORE(EPIT_CR, REG32_LOAD(EPIT_CR) & (~EPIT_CR_OCIEN)); /* Turn off the interrupt */
+	REG32_STORE(EPIT_CR, REG32_LOAD(EPIT_CR) & (~EPIT_CR_EN)); /* disable the timer*/
+	REG32_STORE(EPIT_CR, REG32_LOAD(EPIT_CR) | EPIT_CR_CLKSRC); /* Select the clock source */
+	REG32_STORE(EPIT_SR, EPIT_SR_CLEAR); /* Clear the status register */
+	REG32_STORE(EPIT_CR, REG32_LOAD(EPIT_CR) | EPIT_CR_ENMOD); /* Set enable mode = 1 */
+	REG32_STORE(EPIT_CR, REG32_LOAD(EPIT_CR) | EPIT_CR_EN); /* enable the timer */  
+	REG32_STORE(EPIT_CR, REG32_LOAD(EPIT_CR) | EPIT_CR_OCIEN); /* Enable Interrupts */
+
+	/* setting the load and compare registers */
+	REG32_STORE(EPIT_CR, REG32_LOAD(EPIT_CR) | EPIT_CR_RLD | EPIT_CR_IOVW | EPIT_CR_PRESCLR); /* Set reload and overrite */
+	REG32_STORE(EPIT_LR, 0x3e8); /* Load regiter set to 1000 */
```

Comment:
Should it depend on EPIT_TARGET_HZ value?

Reply:
We can set EPIT_TARGET_HZ to 2000, but this line will be unchanged.

Reply:
Yes, I hard-coded it just for testing and getting review. I will change it, load value will depend on EPIT_TARGET_HZ and EPIT_CR_PRESCLR.